File: mRemoteV1/Tree/Tree.Node.vb
Imports mRemote3G.App
Imports mRemote3G.Connection
Imports mRemote3G.Connection.PuttySession
Imports mRemote3G.Images
Imports mRemote3G.Messages
Imports mRemote3G.Root.PuttySessions

Namespace Tree
    Public Class Node
        Public Enum Type
            None = 0
            Root = 1
            Container = 2
            Connection = 3
            PuttyRoot = 4
            PuttySession = 5
        End Enum

        Private Shared _TreeView As TreeView

        Public Shared Property TreeView As TreeView
            Get
                Return _TreeView
            End Get
            Set
                _TreeView = value
            End Set
        End Property

        Public Shared Property SelectedNode As TreeNode
            Get
                Return _TreeView.SelectedNode
            End Get
            Set
                treeNodeToBeSelected = value
                SelectNode()
            End Set
        End Property

        Private Shared treeNodeToBeSelected As TreeNode

        Private Delegate Sub SelectNodeCB()

        Private Shared Sub SelectNode()
            If _TreeView.InvokeRequired = True Then
                Dim d As New SelectNodeCB(AddressOf SelectNode)
                _TreeView.Invoke(d)
            Else
                _TreeView.SelectedNode = treeNodeToBeSelected
            End If
        End Sub


        Public Shared Function GetConstantID(node As TreeNode) As String
            Select Case GetNodeType(node)
                Case Type.Connection
                    Return TryCast(node.Tag, Info).ConstantID
                Case Type.Container
                    Return TryCast(node.Tag, Container.Info).ConnectionInfo.ConstantID
            End Select

            Return Nothing
        End Function

        Public Shared Function GetNodeFromPositionID(id As Integer) As TreeNode
            For Each conI As Info In Runtime.ConnectionList
                If conI.PositionID = id Then
                    If conI.IsContainer Then
                        Return TryCast(conI.Parent, Container.Info).TreeNode
                    Else
                        Return conI.TreeNode
                    End If
                End If
            Next

            Return Nothing
        End Function

        Public Shared Function GetNodeFromConstantID(id As String) As TreeNode
            For Each conI As Info In Runtime.ConnectionList
                If conI.ConstantID = id Then
                    If conI.IsContainer Then
                        Return TryCast(conI.Parent, Container.Info).TreeNode
                    Else
                        Return conI.TreeNode
                    End If
                End If
            Next

            Return Nothing
        End Function

        Public Shared Function GetNodeType(treeNode As TreeNode) As Type
            Try
                If treeNode Is Nothing Then
                    Return Type.NONE
                End If

                If treeNode.Tag Is Nothing Then
                    Return Type.NONE
                End If

                If TypeOf treeNode.Tag Is PuttySessionsInfo Then
                    Return Type.PuttyRoot
                ElseIf TypeOf treeNode.Tag Is Root.Info Then
                    Return Type.Root
                ElseIf TypeOf treeNode.Tag Is Container.Info Then
                    Return Type.Container
                ElseIf TypeOf treeNode.Tag Is PuttyInfo Then
                    Return Type.PuttySession
                ElseIf TypeOf treeNode.Tag Is Info Then
                    Return Type.Connection
                End If
            Catch ex As Exception
                Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                    "Couldn't get node type" & vbNewLine & ex.ToString(), True)
            End Try

            Return Type.NONE
        End Function

        Public Shared Function GetNodeTypeFromString(str As String) As Type
            Try
                Select Case LCase(str)
                    Case "root"
                        Return Type.Root
                    Case "container"
                        Return Type.Container
                    Case "connection"
                        Return Type.Connection
                End Select
            Catch ex As Exception
                Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                    "Couldn't get node type from string" & vbNewLine & ex.ToString(),
                                                    True)
            End Try

            Return Type.NONE
        End Function

        Public Shared Function Find(treeNode As TreeNode, searchFor As String) As TreeNode
            Dim tmpNode As TreeNode

            Try
                If InStr(LCase(treeNode.Text), LCase(searchFor)) > 0 Then
                    Return treeNode
                Else
                    For Each childNode As TreeNode In treeNode.Nodes
                        tmpNode = Find(childNode, searchFor)
                        If Not tmpNode Is Nothing Then
                            Return tmpNode
                        End If
                    Next
                End If
            Catch ex As Exception
                Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                    "Find node failed" & vbNewLine & ex.ToString(), True)
            End Try

            Return Nothing
        End Function

        Public Shared Function Find(treeNode As TreeNode, conInfo As Info) As TreeNode
            Dim tmpNode As TreeNode

            Try
                If treeNode.Tag Is conInfo Then
                    Return treeNode
                Else
                    For Each childNode As TreeNode In treeNode.Nodes
                        tmpNode = Find(childNode, conInfo)
                        If Not tmpNode Is Nothing Then
                            Return tmpNode
                        End If
                    Next
                End If
            Catch ex As Exception
                Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                    "Find node failed" & vbNewLine & ex.ToString(), True)
            End Try

            Return Nothing
        End Function

        Public Shared Function IsEmpty(treeNode As TreeNode) As Boolean
            Try
                If treeNode.Nodes.Count <= 0 Then
                    Return False
                End If
            Catch ex As Exception
                Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                    "IsEmpty (Tree.Node) failed" & vbNewLine & ex.ToString(), True)
            End Try

            Return True
        End Function


        Public Shared Function AddNode(nodeType As Type, Optional ByVal name As String = Nothing) As TreeNode
            Try
                Dim treeNode As New TreeNode
                Dim defaultName = ""

                Select Case nodeType
                    Case Type.Connection, Type.PuttySession
                        defaultName = Language.Language.strNewConnection
                        treeNode.ImageIndex = Enums.TreeImage.ConnectionClosed
                        treeNode.SelectedImageIndex = Enums.TreeImage.ConnectionClosed
                    Case Type.Container
                        defaultName = Language.Language.strNewFolder
                        treeNode.ImageIndex = Enums.TreeImage.Container
                        treeNode.SelectedImageIndex = Enums.TreeImage.Container
                    Case Type.Root
                        defaultName = Language.Language.strNewRoot
                        treeNode.ImageIndex = Enums.TreeImage.Root
                        treeNode.SelectedImageIndex = Enums.TreeImage.Root
                End Select

                If Not String.IsNullOrEmpty(name) Then
                    treeNode.Name = name
                Else
                    treeNode.Name = defaultName
                End If
                treeNode.Text = treeNode.Name

                Return treeNode
            Catch ex As Exception
                Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg, "AddNode failed" & vbNewLine & ex.ToString(),
                                                    True)
            End Try

            Return Nothing
        End Function

        Public Shared Sub CloneNode(oldTreeNode As TreeNode, Optional ByVal parentNode As TreeNode = Nothing)
            Try
                If GetNodeType(oldTreeNode) = Type.Connection Then
                    Dim oldConnectionInfo = DirectCast(oldTreeNode.Tag, Info)

                    Dim newConnectionInfo As Info = oldConnectionInfo.Copy()
                    Dim newInheritance As Info.Inheritance = oldConnectionInfo.Inherit.Copy()
                    newInheritance.Parent = newConnectionInfo
                    newConnectionInfo.Inherit = newInheritance

                    Runtime.ConnectionList.Add(newConnectionInfo)

                    Dim newTreeNode As New TreeNode(newConnectionInfo.Name)
                    newTreeNode.Tag = newConnectionInfo
                    newTreeNode.ImageIndex = Enums.TreeImage.ConnectionClosed
                    newTreeNode.SelectedImageIndex = Enums.TreeImage.ConnectionClosed

                    newConnectionInfo.TreeNode = newTreeNode

                    If parentNode Is Nothing Then
                        oldTreeNode.Parent.Nodes.Insert(oldTreeNode.Index + 1, newTreeNode)
                        TreeView.SelectedNode = newTreeNode
                    Else
                        Dim parentContainerInfo = TryCast(parentNode.Tag, Container.Info)
                        If parentContainerInfo IsNot Nothing Then
                            newConnectionInfo.Parent = parentContainerInfo
                        End If
                        parentNode.Nodes.Add(newTreeNode)
                    End If
                ElseIf GetNodeType(oldTreeNode) = Type.Container Then
                    Dim oldContainerInfo = DirectCast(oldTreeNode.Tag, Container.Info)

                    Dim newContainerInfo As Container.Info = oldContainerInfo.Copy()
                    Dim newConnectionInfo As Info = oldContainerInfo.ConnectionInfo.Copy()
                    newContainerInfo.ConnectionInfo = newConnectionInfo

                    Dim newTreeNode As New TreeNode(newContainerInfo.Name)
                    newTreeNode.Tag = newContainerInfo
                    newTreeNode.ImageIndex = Enums.TreeImage.Container
                    newTreeNode.SelectedImageIndex = Enums.TreeImage.Container
                    newContainerInfo.ConnectionInfo.Parent = newContainerInfo

                    Runtime.ContainerList.Add(newContainerInfo)

                    If parentNode Is Nothing Then
                        oldTreeNode.Parent.Nodes.Insert(oldTreeNode.Index + 1, newTreeNode)
                        TreeView.SelectedNode = newTreeNode
                    Else
                        parentNode.Nodes.Add(newTreeNode)
                    End If

                    For Each childTreeNode As TreeNode In oldTreeNode.Nodes
                        CloneNode(childTreeNode, newTreeNode)
                    Next

                    newTreeNode.Expand()
                End If
            Catch ex As Exception
                Runtime.MessageCollector.AddMessage(MessageClass.WarningMsg,
                                                    String.Format(Language.Language.strErrorCloneNodeFailed,
                                                                  ex.ToString()))
            End Try
        End Sub

        Public Shared Sub SetNodeImage(treeNode As TreeNode, Img As Enums.TreeImage)
            SetNodeImageIndex(treeNode, Img)
        End Sub

        Private Delegate Sub SetNodeImageIndexDelegate(treeNode As TreeNode, imageIndex As Integer)

        Private Shared Sub SetNodeImageIndex(treeNode As TreeNode, imageIndex As Integer)
            If treeNode Is Nothing OrElse treeNode.TreeView Is Nothing Then Return
            If treeNode.TreeView.InvokeRequired Then
                treeNode.TreeView.Invoke(New SetNodeImageIndexDelegate(AddressOf SetNodeImageIndex),
                                         New Object() {treeNode, imageIndex})
                Return
            End If

            treeNode.ImageIndex = imageIndex
            treeNode.SelectedImageIndex = imageIndex
        End Sub

        Public Shared Sub SetNodeToolTip(e As MouseEventArgs, tTip As ToolTip)
            Try
                If My.Settings.ShowDescriptionTooltipsInTree Then
                    'Find the node under the mouse.
                    Static old_node As TreeNode
                    Dim new_node As TreeNode = _TreeView.GetNodeAt(e.X, e.Y)
                    If new_node Is old_node Then Exit Sub
                    old_node = new_node

                    'See if we have a node.
                    If old_node Is Nothing Then
                        tTip.SetToolTip(_TreeView, "")
                    Else
                        'Get this node's object data.
                        If GetNodeType(old_node) = Type.Connection Then
                            tTip.SetToolTip(_TreeView, TryCast(old_node.Tag, Info).Description)
                        End If
                    End If
                End If
            Catch ex As Exception
                Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                    "SetNodeToolTip failed" & vbNewLine & ex.ToString(), True)
            End Try
        End Sub


        Public Shared Sub DeleteSelectedNode()
            Try
                If SelectedNode Is Nothing Then Return

                Select Case GetNodeType(SelectedNode)
                    Case Type.Root
                        Runtime.MessageCollector.AddMessage(MessageClass.WarningMsg, "The root item cannot be deleted!")
                    Case Type.Container
                        If IsEmpty(SelectedNode) = False Then
                            If _
                                MsgBox(String.Format(Language.Language.strConfirmDeleteNodeFolder, SelectedNode.Text),
                                       MsgBoxStyle.YesNo Or MsgBoxStyle.Question) = MsgBoxResult.Yes Then
                                SelectedNode.Remove()
                            End If
                        Else
                            If _
                                MsgBox(
                                    String.Format(Language.Language.strConfirmDeleteNodeFolderNotEmpty,
                                                  SelectedNode.Text), MsgBoxStyle.YesNo Or MsgBoxStyle.Question) =
                                MsgBoxResult.Yes Then
                                For Each tNode As TreeNode In SelectedNode.Nodes
                                    tNode.Remove()
                                Next
                                SelectedNode.Remove()
                            End If
                        End If
                    Case Type.Connection
                        If _
                            MsgBox(String.Format(Language.Language.strConfirmDeleteNodeConnection, SelectedNode.Text),
                                   MsgBoxStyle.YesNo Or MsgBoxStyle.Question) = MsgBoxResult.Yes Then
                            SelectedNode.Remove()
                        End If
                    Case Else
                        Runtime.MessageCollector.AddMessage(MessageClass.WarningMsg,
                                                            "Tree item type is unknown so it cannot be deleted!")
                End Select
            Catch ex As Exception
                Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                    "Deleting selected node failed" & vbNewLine & ex.ToString(), True)
            End Try
        End Sub

        Public Shared Sub StartRenameSelectedNode()
            If SelectedNode IsNot Nothing Then SelectedNode.BeginEdit()
        End Sub

        Public Shared Sub FinishRenameSelectedNode(newName As String)
            If newName Is Nothing Then Return

            If newName.Length > 0 Then
                SelectedNode.Tag.Name = newName

                If My.Settings.SetHostnameLikeDisplayName Then
                    Dim connectionInfo = TryCast(SelectedNode.Tag, Info)
                    If (connectionInfo IsNot Nothing) Then
                        connectionInfo.Hostname = newName
                    End If
                End If
            End If
        End Sub

        Public Shared Sub MoveNodeUp()
            Try
                If SelectedNode IsNot Nothing Then
                    If Not (SelectedNode.PrevNode Is Nothing) Then
                        TreeView.BeginUpdate()
                        TreeView.Sorted = False

                        Dim newNode As TreeNode = SelectedNode.Clone
                        SelectedNode.Parent.Nodes.Insert(SelectedNode.Index - 1, newNode)
                        SelectedNode.Remove()
                        SelectedNode = newNode

                        TreeView.EndUpdate()
                    End If
                End If
            Catch ex As Exception
                Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                    "MoveNodeUp failed" & vbNewLine & ex.ToString(), True)
            End Try
        End Sub

        Public Shared Sub MoveNodeDown()
            Try
                If SelectedNode IsNot Nothing Then
                    If Not (SelectedNode.NextNode Is Nothing) Then
                        TreeView.BeginUpdate()
                        TreeView.Sorted = False

                        Dim newNode As TreeNode = SelectedNode.Clone
                        SelectedNode.Parent.Nodes.Insert(SelectedNode.Index + 2, newNode)
                        SelectedNode.Remove()
                        SelectedNode = newNode

                        TreeView.EndUpdate()
                    End If
                End If
            Catch ex As Exception
                Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                    "MoveNodeDown failed" & vbNewLine & ex.ToString(), True)
            End Try
        End Sub

        Public Shared Sub ExpandAllNodes()
            TreeView.BeginUpdate()
            TreeView.ExpandAll()
            TreeView.EndUpdate()
        End Sub

        Public Shared Sub CollapseAllNodes()
            TreeView.BeginUpdate()
            For Each treeNode As TreeNode In TreeView.Nodes(0).Nodes
                treeNode.Collapse(False)
            Next
            TreeView.EndUpdate()
        End Sub

        Public Shared Sub Sort(treeNode As TreeNode, sorting As SortOrder)
            If TreeView Is Nothing Then Return

            TreeView.BeginUpdate()

            If treeNode Is Nothing Then
                If TreeView.Nodes.Count > 0 Then
                    treeNode = TreeView.Nodes.Item(0)
                Else
                    Return
                End If
            ElseIf GetNodeType(treeNode) = Type.Connection Then
                treeNode = treeNode.Parent
                If treeNode Is Nothing Then Return
            End If

            Sort(treeNode, New Tools.Controls.TreeNodeSorter(sorting))

            TreeView.EndUpdate()
        End Sub

        ' Adapted from http://www.codeproject.com/Tips/252234/ASP-NET-TreeView-Sort
        Private Shared Sub Sort(treeNode As TreeNode, nodeSorter As Tools.Controls.TreeNodeSorter)
            For Each childNode As TreeNode In treeNode.Nodes
                Sort(childNode, nodeSorter)
            Next

            Try
                Dim sortedNodes As New List(Of TreeNode)
                Dim currentNode As TreeNode = Nothing
                While (treeNode.Nodes.Count > 0)
                    For Each childNode As TreeNode In treeNode.Nodes
                        If (currentNode Is Nothing OrElse nodeSorter.Compare(childNode, currentNode) < 0) Then
                            currentNode = childNode
                        End If
                    Next
                    treeNode.Nodes.Remove(currentNode)
                    sortedNodes.Add(currentNode)
                    currentNode = Nothing
                End While

                For Each childNode As TreeNode In sortedNodes
                    treeNode.Nodes.Add(childNode)
                Next
            Catch ex As Exception
                Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                    "Sort nodes failed" & vbNewLine & ex.ToString(), True)
            End Try
        End Sub

        Private Delegate Sub ResetTreeDelegate()

        Public Shared Sub ResetTree()
            If TreeView.InvokeRequired Then
                Dim resetTreeDelegate As New ResetTreeDelegate(AddressOf ResetTree)
                Runtime.Windows.treeForm.Invoke(resetTreeDelegate)
            Else
                TreeView.BeginUpdate()
                TreeView.Nodes.Clear()
                TreeView.Nodes.Add(Language.Language.strConnections)
                TreeView.EndUpdate()
            End If
        End Sub

        Private Sub New()
        End Sub
    End Class
End Namespace

File: mRemoteV1/CS/Tree/Tree.Node.cs
// VBConversions Note: VB project level imports
using System.Collections.Generic;
using System;
using AxWFICALib;
using System.Drawing;
using System.Diagnostics;
using System.Data;
using AxMSTSCLib;
using Microsoft.VisualBasic;
using System.Collections;
using System.Windows.Forms;
// End of VB project level imports

//using mRemoteNG.App.Runtime;
using System.DirectoryServices;


namespace mRemoteNG.Tree
{
	public class Node
		{
			public enum Type
			{
				None = 0,
				Root = 1,
				Container = 2,
				Connection = 3,
				PuttyRoot = 4,
				PuttySession = 5
			}
			
			private static TreeView _TreeView;
public static TreeView TreeView
			{
				get
				{
					return _TreeView;
				}
				set
				{
					_TreeView = value;
				}
			}
			
public static TreeNode SelectedNode
			{
				get
				{
					return _TreeView.SelectedNode;
				}
				set
				{
					treeNodeToBeSelected = value;
					SelectNode();
				}
			}
			
			private static TreeNode treeNodeToBeSelected;
			private delegate void SelectNodeCB();
			private static void SelectNode()
			{
				if (_TreeView.InvokeRequired == true)
				{
					SelectNodeCB d = new SelectNodeCB(SelectNode);
					_TreeView.Invoke(d);
				}
				else
				{
					_TreeView.SelectedNode = treeNodeToBeSelected;
				}
			}
			
			
			public static string GetConstantID(TreeNode node)
			{
				if (GetNodeType(node) == Type.Connection)
				{
					return (node.Tag as mRemoteNG.Connection.Info).ConstantID;
				}
				else if (GetNodeType(node) == Type.Container)
				{
					return (node.Tag as mRemoteNG.Container.Info).ConnectionInfo.ConstantID;
				}
				
				return null;
			}
			
			public static TreeNode GetNodeFromPositionID(int id)
			{
				foreach (Connection.Info conI in ConnectionList)
				{
					if (conI.PositionID == id)
					{
						if (conI.IsContainer)
						{
							return (conI.Parent as Container.Info).TreeNode;
						}
						else
						{
							return conI.TreeNode;
						}
					}
				}
				
				return null;
			}
			
			public static TreeNode GetNodeFromConstantID(string id)
			{
				foreach (Connection.Info conI in ConnectionList)
				{
					if (conI.ConstantID == id)
					{
						if (conI.IsContainer)
						{
							return (conI.Parent as Container.Info).TreeNode;
						}
						else
						{
							return conI.TreeNode;
						}
					}
				}
				
				return null;
			}
			
			public static Tree.Node.Type GetNodeType(TreeNode treeNode)
			{
				try
				{
					if (treeNode == null)
					{
						return Type.None;
					}
					
					if (treeNode.Tag == null)
					{
						return Type.None;
					}
					
					if (treeNode.Tag is Root.PuttySessions.Info)
					{
						return Type.PuttyRoot;
					}
					else if (treeNode.Tag is Root.Info)
					{
						return Type.Root;
					}
					else if (treeNode.Tag is Container.Info)
					{
						return Type.Container;
					}
					else if (treeNode.Tag is Connection.PuttySession.Info)
					{
						return Type.PuttySession;
					}
					else if (treeNode.Tag is Connection.Info)
					{
						return Type.Connection;
					}
				}
				catch (Exception ex)
				{
					MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "Couldn\'t get node type" + Constants.vbNewLine + ex.Message, true);
				}
				
				return Type.None;
			}
			
			public static Tree.Node.Type GetNodeTypeFromString(string str)
			{
				try
				{
					switch (str.ToLower())
					{
						case "root":
							return Type.Root;
						case "container":
							return Type.Container;
						case "connection":
							return Type.Connection;
					}
				}
				catch (Exception ex)
				{
					MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "Couldn\'t get node type from string" + Constants.vbNewLine + ex.Message, true);
				}
				
				return Type.None;
			}
			
			public static TreeNode Find(TreeNode treeNode, string searchFor)
			{
				TreeNode tmpNode = default(TreeNode);
				
				try
				{
					if ((treeNode.Text.ToLower()).IndexOf(searchFor.ToLower()) + 1 > 0)
					{
						return treeNode;
					}
					else
					{
						foreach (TreeNode childNode in treeNode.Nodes)
						{
							tmpNode = Find(childNode, searchFor);
							if (!(tmpNode == null))
							{
								return tmpNode;
							}
						}
					}
				}
				catch (Exception ex)
				{
					MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "Find node failed" + Constants.vbNewLine + ex.Message, true);
				}
				
				return null;
			}
			
			public static TreeNode Find(TreeNode treeNode, Connection.Info conInfo)
			{
				TreeNode tmpNode = default(TreeNode);
				
				try
				{
					if (treeNode.Tag == conInfo)
					{
						return treeNode;
					}
					else
					{
						foreach (TreeNode childNode in treeNode.Nodes)
						{
							tmpNode = Find(childNode, conInfo);
							if (!(tmpNode == null))
							{
								return tmpNode;
							}
						}
					}
				}
				catch (Exception ex)
				{
					MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "Find node failed" + Constants.vbNewLine + ex.Message, true);
				}
				
				return null;
			}
			
			public static bool IsEmpty(TreeNode treeNode)
			{
				try
				{
					if (treeNode.Nodes.Count <= 0)
					{
						return false;
					}
				}
				catch (Exception ex)
				{
					MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "IsEmpty (Tree.Node) failed" + Constants.vbNewLine + ex.Message, true);
				}
				
				return true;
			}
			
			
			
			public static TreeNode AddNode(Type nodeType, string name = null)
			{
				try
				{
					TreeNode treeNode = new TreeNode();
					string defaultName = "";
					
					switch (nodeType)
					{
						case Type.Connection:
						case Type.PuttySession:
							defaultName = My.Language.strNewConnection;
							treeNode.ImageIndex = Images.Enums.TreeImage.ConnectionClosed;
							treeNode.SelectedImageIndex = Images.Enums.TreeImage.ConnectionClosed;
							break;
						case Type.Container:
							defaultName = My.Language.strNewFolder;
							treeNode.ImageIndex = Images.Enums.TreeImage.Container;
							treeNode.SelectedImageIndex = Images.Enums.TreeImage.Container;
							break;
						case Type.Root:
							defaultName = My.Language.strNewRoot;
							treeNode.ImageIndex = Images.Enums.TreeImage.Root;
							treeNode.SelectedImageIndex = Images.Enums.TreeImage.Root;
							break;
					}
					
					if (!string.IsNullOrEmpty(name))
					{
						treeNode.Name = name;
					}
					else
					{
						treeNode.Name = defaultName;
					}
					treeNode.Text = treeNode.Name;
					
					return treeNode;
				}
				catch (Exception ex)
				{
					MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "AddNode failed" + Constants.vbNewLine + ex.Message, true);
				}
				
				return null;
			}
			
			public static void CloneNode(TreeNode oldTreeNode, TreeNode parentNode = null)
			{
				try
				{
					if (GetNodeType(oldTreeNode) == Type.Connection)
					{
						Connection.Info oldConnectionInfo = (Connection.Info) oldTreeNode.Tag;
						
						Connection.Info newConnectionInfo = oldConnectionInfo.Copy();
						Connection.Info.Inheritance newInheritance = oldConnectionInfo.Inherit.Copy();
						newInheritance.Parent = newConnectionInfo;
						newConnectionInfo.Inherit = newInheritance;
						
						ConnectionList.Add(newConnectionInfo);
						
						TreeNode newTreeNode = new TreeNode(newConnectionInfo.Name);
						newTreeNode.Tag = newConnectionInfo;
						newTreeNode.ImageIndex = Images.Enums.TreeImage.ConnectionClosed;
						newTreeNode.SelectedImageIndex = Images.Enums.TreeImage.ConnectionClosed;
						
						newConnectionInfo.TreeNode = newTreeNode;
						
						if (parentNode == null)
						{
							oldTreeNode.Parent.Nodes.Insert(oldTreeNode.Index + 1, newTreeNode);
							TreeView.SelectedNode = newTreeNode;
						}
						else
						{
							Container.Info parentContainerInfo = parentNode.Tag as Container.Info;
							if (parentContainerInfo != null)
							{
								newConnectionInfo.Parent = parentContainerInfo;
							}
							parentNode.Nodes.Add(newTreeNode);
						}
					}
					else if (GetNodeType(oldTreeNode) == Type.Container)
					{
						Container.Info oldContainerInfo = (Container.Info) oldTreeNode.Tag;
						
						Container.Info newContainerInfo = oldContainerInfo.Copy();
						Connection.Info newConnectionInfo = oldContainerInfo.ConnectionInfo.Copy();
						newContainerInfo.ConnectionInfo = newConnectionInfo;
						
						TreeNode newTreeNode = new TreeNode(newContainerInfo.Name);
						newTreeNode.Tag = newContainerInfo;
						newTreeNode.ImageIndex = Images.Enums.TreeImage.Container;
						newTreeNode.SelectedImageIndex = Images.Enums.TreeImage.Container;
						newContainerInfo.ConnectionInfo.Parent = newContainerInfo;
						
						ContainerList.Add(newContainerInfo);
						
						if (parentNode == null)
						{
							oldTreeNode.Parent.Nodes.Insert(oldTreeNode.Index + 1, newTreeNode);
							TreeView.SelectedNode = newTreeNode;
						}
						else
						{
							parentNode.Nodes.Add(newTreeNode);
						}
						
						foreach (TreeNode childTreeNode in oldTreeNode.Nodes)
						{
							CloneNode(childTreeNode, newTreeNode);
						}
						
						newTreeNode.Expand();
					}
				}
				catch (Exception ex)
				{
					MessageCollector.AddMessage(Messages.MessageClass.WarningMsg, string.Format(My.Language.strErrorCloneNodeFailed, ex.Message));
				}
			}
			
			public static void SetNodeImage(TreeNode treeNode, Images.Enums.TreeImage Img)
			{
				SetNodeImageIndex(treeNode, Img);
			}
			
			private delegate void SetNodeImageIndexDelegate(TreeNode treeNode, int imageIndex);
			private static void SetNodeImageIndex(TreeNode treeNode, int imageIndex)
			{
				if (treeNode == null || treeNode.TreeView == null)
				{
					return ;
				}
				if (treeNode.TreeView.InvokeRequired)
				{
					treeNode.TreeView.Invoke(new SetNodeImageIndexDelegate(SetNodeImageIndex), new object[] {treeNode, imageIndex});
					return ;
				}
				
				treeNode.ImageIndex = imageIndex;
				treeNode.SelectedImageIndex = imageIndex;
			}
			
			// VBConversions Note: Former VB static variables moved to class level because they aren't supported in C#.
			static TreeNode SetNodeToolTip_old_node = default(TreeNode);
			
			public static void SetNodeToolTip(MouseEventArgs e, ToolTip tTip)
			{
				try
				{
					if (My.Settings.Default.ShowDescriptionTooltipsInTree)
					{
						//Find the node under the mouse.
						// static TreeNode old_node = default(TreeNode); VBConversions Note: Static variable moved to class level and renamed SetNodeToolTip_old_node. Local static variables are not supported in C#.
						TreeNode new_node = _TreeView.GetNodeAt(e.X, e.Y);
						if (new_node == SetNodeToolTip_old_node)
						{
							return;
						}
						SetNodeToolTip_old_node = new_node;
						
						//See if we have a node.
						if (SetNodeToolTip_old_node == null)
						{
							tTip.SetToolTip(_TreeView, "");
						}
						else
						{
							//Get this node's object data.
							if (GetNodeType(SetNodeToolTip_old_node) == Type.Connection)
							{
								tTip.SetToolTip(_TreeView, (SetNodeToolTip_old_node.Tag as mRemoteNG.Connection.Info).Description);
							}
						}
					}
				}
				catch (Exception ex)
				{
					MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "SetNodeToolTip failed" + Constants.vbNewLine + ex.Message, true);
				}
			}
			
			
			public static void DeleteSelectedNode()
			{
				try
				{
					if (SelectedNode == null)
					{
						return ;
					}
					
					if (Tree.Node.GetNodeType(SelectedNode) == Type.Root)
					{
						MessageCollector.AddMessage(Messages.MessageClass.WarningMsg, "The root item cannot be deleted!");
					}
					else if (Tree.Node.GetNodeType(SelectedNode) == Type.Container)
					{
						if (Tree.Node.IsEmpty(SelectedNode) == false)
						{
							if (Interaction.MsgBox(string.Format(My.Language.strConfirmDeleteNodeFolder, SelectedNode.Text), (Microsoft.VisualBasic.MsgBoxStyle) (MsgBoxStyle.YesNo | MsgBoxStyle.Question), null) == MsgBoxResult.Yes)
							{
								SelectedNode.Remove();
							}
						}
						else
						{
							if (Interaction.MsgBox(string.Format(My.Language.strConfirmDeleteNodeFolderNotEmpty, SelectedNode.Text), (Microsoft.VisualBasic.MsgBoxStyle) (MsgBoxStyle.YesNo | MsgBoxStyle.Question), null) == MsgBoxResult.Yes)
							{
								foreach (TreeNode tNode in SelectedNode.Nodes)
								{
									tNode.Remove();
								}
								SelectedNode.Remove();
							}
						}
					}
					else if (Tree.Node.GetNodeType(SelectedNode) == Type.Connection)
					{
						if (Interaction.MsgBox(string.Format(My.Language.strConfirmDeleteNodeConnection, SelectedNode.Text), (Microsoft.VisualBasic.MsgBoxStyle) (MsgBoxStyle.YesNo | MsgBoxStyle.Question), null) == MsgBoxResult.Yes)
						{
							SelectedNode.Remove();
						}
					}
					else
					{
						MessageCollector.AddMessage(Messages.MessageClass.WarningMsg, "Tree item type is unknown so it cannot be deleted!");
					}
				}
				catch (Exception ex)
				{
					MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "Deleting selected node failed" + Constants.vbNewLine + ex.Message, true);
				}
			}
			
			public static void StartRenameSelectedNode()
			{
				if (SelectedNode != null)
				{
					SelectedNode.BeginEdit();
				}
			}
			
			public static void FinishRenameSelectedNode(string newName)
			{
				if (newName == null)
				{
					return ;
				}
				
				if (newName.Length > 0)
				{
					SelectedNode.Tag.Name = newName;
					
					if (My.Settings.Default.SetHostnameLikeDisplayName)
					{
						Connection.Info connectionInfo = SelectedNode.Tag as Connection.Info;
						if (connectionInfo != null)
						{
							connectionInfo.Hostname = newName;
						}
					}
				}
			}
			
			public static void MoveNodeUp()
			{
				try
				{
					if (SelectedNode != null)
					{
						if (!(SelectedNode.PrevNode == null))
						{
							TreeView.BeginUpdate();
							TreeView.Sorted = false;
							
							TreeNode newNode = SelectedNode.Clone();
							SelectedNode.Parent.Nodes.Insert(SelectedNode.Index - 1, newNode);
							SelectedNode.Remove();
							SelectedNode = newNode;
							
							TreeView.EndUpdate();
						}
					}
				}
				catch (Exception ex)
				{
					MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "MoveNodeUp failed" + Constants.vbNewLine + ex.Message, true);
				}
			}
			
			public static void MoveNodeDown()
			{
				try
				{
					if (SelectedNode != null)
					{
						if (!(SelectedNode.NextNode == null))
						{
							TreeView.BeginUpdate();
							TreeView.Sorted = false;
							
							TreeNode newNode = SelectedNode.Clone();
							SelectedNode.Parent.Nodes.Insert(SelectedNode.Index + 2, newNode);
							SelectedNode.Remove();
							SelectedNode = newNode;
							
							TreeView.EndUpdate();
						}
					}
				}
				catch (Exception ex)
				{
					MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "MoveNodeDown failed" + Constants.vbNewLine + ex.Message, true);
				}
			}
			
			public static void ExpandAllNodes()
			{
				TreeView.BeginUpdate();
				TreeView.ExpandAll();
				TreeView.EndUpdate();
			}
			
			public static void CollapseAllNodes()
			{
				TreeView.BeginUpdate();
				foreach (TreeNode treeNode in TreeView.Nodes[0].Nodes)
				{
					treeNode.Collapse(false);
				}
				TreeView.EndUpdate();
			}
			
			public static void Sort(TreeNode treeNode, System.Windows.Forms.SortOrder sorting)
			{
				if (TreeView == null)
				{
					return ;
				}
				
				TreeView.BeginUpdate();
				
				if (treeNode == null)
				{
					if (TreeView.Nodes.Count > 0)
					{
						treeNode = TreeView.Nodes[0];
					}
					else
					{
						return ;
					}
				}
				else if (GetNodeType(treeNode) == Type.Connection)
				{
					treeNode = treeNode.Parent;
					if (treeNode == null)
					{
						return ;
					}
				}
				
				Sort(treeNode, new Tools.Controls.TreeNodeSorter(sorting));
				
				TreeView.EndUpdate();
			}
			
			// Adapted from http://www.codeproject.com/Tips/252234/ASP-NET-TreeView-Sort
			private static void Sort(TreeNode treeNode, Tools.Controls.TreeNodeSorter nodeSorter)
			{
				foreach (TreeNode childNode in treeNode.Nodes)
				{
					Sort(childNode, nodeSorter);
				}
				
				try
				{
					List<TreeNode> sortedNodes = new List<TreeNode>();
					TreeNode currentNode = null;
					while (treeNode.Nodes.Count > 0)
					{
						foreach (TreeNode childNode in treeNode.Nodes)
						{
							if (currentNode == null || nodeSorter.Compare(childNode, currentNode) < 0)
							{
								currentNode = childNode;
							}
						}
						treeNode.Nodes.Remove(currentNode);
						sortedNodes.Add(currentNode);
						currentNode = null;
					}
					
					foreach (TreeNode childNode in sortedNodes)
					{
						treeNode.Nodes.Add(childNode);
					}
				}
				catch (Exception ex)
				{
					MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "Sort nodes failed" + Constants.vbNewLine + ex.Message, true);
				}
			}
			
			private delegate void ResetTreeDelegate();
			public static void ResetTree()
			{
				if (TreeView.InvokeRequired)
				{
					ResetTreeDelegate resetTreeDelegate = new ResetTreeDelegate(ResetTree);
					Windows.treeForm.Invoke(resetTreeDelegate);
				}
				else
				{
					TreeView.BeginUpdate();
					TreeView.Nodes.Clear();
					TreeView.Nodes.Add(My.Language.strConnections);
					TreeView.EndUpdate();
				}
			}
		}
}
